File: golang/801_min_swap/min_swap.go
/*
 * @Date: 2022-10-10
 * @LastEditors: mengsen_wang@163.com
 * @LastEditTime: 2022-10-10
 * @FilePath: /algorithm/801_min_swap/min_swap.go
 */

package main

func minSwap(nums1, nums2 []int) int {
	min := func(a, b int) int {
		if a > b {
			return b
		}
		return a
	}
	n := len(nums1)
	a, b := 0, 1
	for i := 1; i < n; i++ {
		at, bt := a, b
		a, b = n, n
		if nums1[i] > nums1[i-1] && nums2[i] > nums2[i-1] {
			a = min(a, at)
			b = min(b, bt+1)
		}
		if nums1[i] > nums2[i-1] && nums2[i] > nums1[i-1] {
			a = min(a, bt)
			b = min(b, at+1)
		}
	}
	return min(a, b)
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}
	{
		nums1 := []int{1, 3, 5, 4}
		nums2 := []int{1, 2, 3, 7}
		ans := 1
		assert(minSwap(nums1, nums2) == ans)
	}

	{
		nums1 := []int{0, 3, 5, 8, 9}
		nums2 := []int{2, 1, 4, 6, 9}
		ans := 1
		assert(minSwap(nums1, nums2) == ans)
	}
}

File: cpp/801_min_swap/min_swap.cpp
/*
 * @Date: 2022-10-10
 * @LastEditors: mengsen_wang@163.com
 * @LastEditTime: 2022-10-10
 * @FilePath: /algorithm/801_min_swap/min_swap.cpp
 */

#include <cassert>
#include <vector>

using namespace std;

class Solution {
 public:
  int minSwap(vector<int>& nums1, vector<int>& nums2) {
    int n = nums1.size();
    int a = 0, b = 1;
    for (int i = 1; i < n; i++) {
      int at = a, bt = b;
      a = b = n;
      if (nums1[i] > nums1[i - 1] && nums2[i] > nums2[i - 1]) {
        a = min(a, at);
        b = min(b, bt + 1);
      }
      if (nums1[i] > nums2[i - 1] && nums2[i] > nums1[i - 1]) {
        a = min(a, bt);
        b = min(b, at + 1);
      }
    }
    return min(a, b);
  }
};

int main() {
  {
    vector<int> nums1{1, 3, 5, 4};
    vector<int> nums2{1, 2, 3, 7};
    int ans = 1;
    assert(Solution().minSwap(nums1, nums2) == ans);
  }

  {
    vector<int> nums1{0, 3, 5, 8, 9};
    vector<int> nums2{2, 1, 4, 6, 9};
    int ans = 1;
    assert(Solution().minSwap(nums1, nums2) == ans);
  }
}